File: breastCancer_10K.py
import math
import random
import time

# Sigmoid function
def sigmoid(x):
    return 1 / (1 + math.exp(-x))

# Mini-Batch Gradient Descent for Faster Training
def logistic_regression(x, y, learning_rate, epochs, batch_size=500):
    num_samples = len(x)
    num_features = len(x[0])
    weights = [0.0] * num_features

    for _ in range(epochs):
        for i in range(0, num_samples, batch_size):  # Process in chunks
            x_batch = x[i:i + batch_size]
            y_batch = y[i:i + batch_size]

            gradients = [0.0] * num_features
            batch_size_actual = len(x_batch)

            for j in range(batch_size_actual):
                y_pred = sigmoid(sum(weights[k] * x_batch[j][k] for k in range(num_features)))
                error = y_pred - y_batch[j]
                for k in range(num_features):
                    gradients[k] += error * x_batch[j][k]

            # Update weights for the batch
            weights = [weights[k] - (learning_rate * gradients[k] / batch_size_actual) for k in range(num_features)]

    return weights

# Optimized synthetic dataset generation
def generate_data(num_samples=10000, num_features=30):
    return (
        [[random.uniform(0, 1) for _ in range(num_features)] for _ in range(num_samples)],
        [1.0 if sum(sample) > (num_features * 0.5) else 0.0 for sample in [[random.uniform(0, 1) for _ in range(num_features)] for _ in range(num_samples)]]
    )

# Measure execution time and accuracy with statistical precision control
def benchmark(num_samples=10000, num_features=30):
    learning_rate = 0.01
    epochs = 200  # Keep 200 for now
    batch_size = 500  # Mini-Batch Size

    execution_times = []
    accuracies = []

    # Statistical precision control (95% CI with 2.5% precision)
    precision_target = 0.025  # 2.5% precision
    confidence_level = 0.95
    sample_size = 30
    mean_time = 0

    while True:
        for run in range(sample_size):
            x, y = generate_data(num_samples, num_features)
            start_time = time.time()
            weights = logistic_regression(x, y, learning_rate, epochs, batch_size)
            exec_time = time.time() - start_time
            execution_times.append(exec_time)

            # Compute accuracy efficiently
            correct_predictions = sum(
                (sigmoid(sum(weights[j] * x[i][j] for j in range(num_features))) > 0.5) == y[i]
                for i in range(num_samples)
            )
            accuracy = correct_predictions / num_samples
            accuracies.append(accuracy)

        # Calculate mean and standard deviation
        mean_time = sum(execution_times) / len(execution_times)
        std_dev = (sum((x - mean_time) ** 2 for x in execution_times) / len(execution_times)) ** 0.5
        
        # Calculate margin of error for 95% CI using a simplified method (Student's t-distribution assumption)
        margin_of_error = 1.96 * (std_dev / (len(execution_times) ** 0.5))

        # Check if the margin of error is within the target precision
        if margin_of_error / mean_time <= precision_target:
            break

    mean_accuracy = sum(accuracies) / len(accuracies)
    return mean_time, mean_accuracy

if __name__ == "__main__":
    num_samples = 10000  # Default value
    mean_time, mean_accuracy = benchmark(num_samples)
    print(f"Mean Execution Time: {mean_time:.4f} seconds")
    print(f"Mean Accuracy: {mean_accuracy:.4f}")

File: breastCancer_10K.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <chrono>
#include <random>
#include <numeric>
#include <algorithm>
#include <cassert>
#ifdef __EMSCRIPTEN__
#include <emscripten.h>
#else
#include <thread>
#endif

// Sigmoid function for logistic regression
double sigmoid(double x) {
    return 1.0 / (1.0 + std::exp(-x));
}

// Logistic regression function with batch gradient descent
std::vector<double> logistic_regression(const std::vector<std::vector<double>>& x, const std::vector<double>& y, double learning_rate, int epochs) {
    size_t num_features = x[0].size();
    std::vector<double> weights(num_features, 0.0);

    for (int epoch = 0; epoch < epochs; ++epoch) {
        std::vector<double> gradients(num_features, 0.0);

        for (size_t i = 0; i < x.size(); ++i) {
            double y_pred = 0.0;
            for (size_t j = 0; j < num_features; ++j) {
                y_pred += weights[j] * x[i][j];
            }
            y_pred = sigmoid(y_pred);
            double error = y_pred - y[i];

            for (size_t j = 0; j < num_features; ++j) {
                gradients[j] += error * x[i][j];
            }
        }

        for (size_t j = 0; j < num_features; ++j) {
            weights[j] -= learning_rate * gradients[j] / static_cast<double>(x.size());
        }
    }

    return weights;
}

// Function to calculate accuracy
double calculate_accuracy(const std::vector<std::vector<double>>& x, const std::vector<double>& y, const std::vector<double>& weights) {
    int correct = 0;
    for (size_t i = 0; i < x.size(); ++i) {
        double y_pred = 0.0;
        for (size_t j = 0; j < weights.size(); ++j) {
            y_pred += weights[j] * x[i][j];
        }
        y_pred = sigmoid(y_pred);
        if ((y_pred >= 0.5 && y[i] == 1.0) || (y_pred < 0.5 && y[i] == 0.0)) {
            ++correct;
        }
    }
    return static_cast<double>(correct) / x.size() * 100.0;
}

// Generate synthetic dataset
void generate_dataset(std::vector<std::vector<double>>& x, std::vector<double>& y, size_t num_samples, size_t num_features) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<> dis(0.0, 100.0);

    x.resize(num_samples, std::vector<double>(num_features));
    y.resize(num_samples);

    for (size_t i = 0; i < num_samples; ++i) {
        for (size_t j = 0; j < num_features; ++j) {
            x[i][j] = dis(gen);
        }
        y[i] = (x[i][0] + x[i][1] > 50.0) ? 1.0 : 0.0;
    }
}

// Perform Student's t-test for precision control
bool check_precision(const std::vector<double>& times, double precision) {
    double mean = std::accumulate(times.begin(), times.end(), 0.0) / times.size();
    double sq_sum = std::inner_product(times.begin(), times.end(), times.begin(), 0.0);
    double stddev = std::sqrt(sq_sum / times.size() - mean * mean);

    double margin_of_error = 1.96 * (stddev / std::sqrt(times.size()));
    return (margin_of_error / mean) <= precision;
}

int main() {
    size_t num_samples = 10000;  // Adjust this for different data sizes
    size_t num_features = 30;
    double learning_rate = 0.1;
    int epochs = 1000;
    double precision = 0.025;  // 2.5% precision

    // Generate dataset
    std::vector<std::vector<double>> x;
    std::vector<double> y;
    generate_dataset(x, y, num_samples, num_features);

    // Split dataset into 80% training and 20% testing
    size_t train_size = static_cast<size_t>(0.8 * num_samples);
    std::vector<std::vector<double>> x_train(x.begin(), x.begin() + train_size);
    std::vector<double> y_train(y.begin(), y.begin() + train_size);
    std::vector<std::vector<double>> x_test(x.begin() + train_size, x.end());
    std::vector<double> y_test(y.begin() + train_size, y.end());

    std::vector<double> execution_times;
    double total_accuracy = 0.0;

    do {
        auto start_time = std::chrono::high_resolution_clock::now();
        std::vector<double> weights = logistic_regression(x_train, y_train, learning_rate, epochs);
        auto duration = std::chrono::high_resolution_clock::now() - start_time;

        execution_times.push_back(std::chrono::duration_cast<std::chrono::microseconds>(duration).count() / 1e6);

        double accuracy = calculate_accuracy(x_test, y_test, weights);
        total_accuracy += accuracy;

#ifndef __EMSCRIPTEN__
        // Introduce delay to prevent cache effects (native only)
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
#endif

    } while (!check_precision(execution_times, precision));

    double mean_time = std::accumulate(execution_times.begin(), execution_times.end(), 0.0) / execution_times.size();
    double mean_accuracy = total_accuracy / execution_times.size();

    std::cout << "Mean execution time: " << mean_time << " seconds\n";
    std::cout << "Mean accuracy: " << mean_accuracy << " %\n";

    return 0;
}

